### PR TITLE
 Implement PhenomDECO - a BBH waveform to infer compactness

### PR DESCRIPTION
**Pull request by deepchatterjeeligo**
Porting over #240. Apologies Shrobana, in the attempt of making this PR final, I force pushed to your dev branch, and it somehow closed the PR #240

### CODE REVIEW

**Comment by deepchatterjeeligo** [comment]
@Shrobana sorry I am not a 100% sure why GitHub did this. However, all your commits and contributions are present in this PR.

**Comment by deepchatterjeeligo** [comment]
Benchmark compare (A100)

```
$ uv run python benchmarks/compare_benchmarks.py .benchmarks/Linux-CPython-3.12-64bit/0001_baseline.json .benchmarks/Linux-CPython-3.12-64bit/0002_deco.json
                                                                                         
| Benchmark                                    |  Baseline |   Current | Delta             |
|----------------------------------------------|-----------|-----------|-------------------|
| test_sine_gaussian_forward[batch_32]         | 538.37 μs | 541.99 μs | Stable (+0.7%)    |
| test_sine_gaussian_forward[batch_128]        | 553.98 μs | 554.90 μs | Stable (+0.2%)    |
| test_sine_gaussian_forward[batch_512]        |   1.11 ms |   1.12 ms | Stable (+0.8%)    |
| test_ringdown_forward[batch_32]              | 573.14 μs | 578.35 μs | Stable (+0.9%)    |
| test_ringdown_forward[batch_128]             | 605.20 μs | 604.65 μs | Stable (-0.1%)    |
| test_ringdown_forward[batch_512]             |   1.04 ms |   1.04 ms | Stable (+0.1%)    |
| test_taylorf2_forward[batch_32-dur_2]        |   4.60 ms |   4.70 ms | Stable (+2.0%)    |
| test_taylorf2_forward[batch_32-dur_4]        |   4.59 ms |   4.69 ms | Stable (+2.3%)    |
| test_taylorf2_forward[batch_32-dur_8]        |   4.59 ms |   4.69 ms | Stable (+2.2%)    |
| test_taylorf2_forward[batch_128-dur_2]       |   4.63 ms |   4.70 ms | Stable (+1.3%)    |
| test_taylorf2_forward[batch_128-dur_4]       |   4.62 ms |   4.73 ms | Stable (+2.4%)    |
| test_taylorf2_forward[batch_128-dur_8]       |   4.67 ms |   4.77 ms | Stable (+2.1%)    |
| test_taylorf2_forward[batch_512-dur_2]       |   4.71 ms |   4.79 ms | Stable (+1.8%)    |
| test_taylorf2_forward[batch_512-dur_4]       |   5.02 ms |   5.12 ms | Stable (+1.9%)    |
| test_taylorf2_forward[batch_512-dur_8]       |   6.63 ms |   6.74 ms | Stable (+1.6%)    |
| test_phenomd_forward[batch_32-dur_2]         |  34.35 ms |  35.22 ms | Stable (+2.6%)    |
| test_phenomd_forward[batch_32-dur_4]         |  34.13 ms |  35.05 ms | Stable (+2.7%)    |
| test_phenomd_forward[batch_32-dur_8]         |  34.01 ms |  35.07 ms | Stable (+3.1%)    |
| test_phenomd_forward[batch_128-dur_2]        |  34.46 ms |  35.31 ms | Stable (+2.5%)    |
| test_phenomd_forward[batch_128-dur_4]        |  34.44 ms |  35.33 ms | Stable (+2.6%)    |
| test_phenomd_forward[batch_128-dur_8]        |  34.82 ms |  35.50 ms | Stable (+2.0%)    |
| test_phenomd_forward[batch_512-dur_2]        |  34.74 ms |  35.55 ms | Stable (+2.3%)    |
| test_phenomd_forward[batch_512-dur_4]        |  35.99 ms |  36.55 ms | Stable (+1.6%)    |
| test_phenomd_forward[batch_512-dur_8]        |  38.36 ms |  39.31 ms | Stable (+2.5%)    |
| test_phenompv2_forward[batch_32-dur_2]       |  39.57 ms |  40.87 ms | Stable (+3.3%)    |
| test_phenompv2_forward[batch_32-dur_4]       |  39.57 ms |  40.85 ms | Stable (+3.2%)    |
| test_phenompv2_forward[batch_32-dur_8]       |  39.41 ms |  40.86 ms | Stable (+3.7%)    |
| test_phenompv2_forward[batch_128-dur_2]      |  39.90 ms |  41.05 ms | Stable (+2.9%)    |
| test_phenompv2_forward[batch_128-dur_4]      |  39.86 ms |  41.04 ms | Stable (+2.9%)    |
| test_phenompv2_forward[batch_128-dur_8]      |  40.79 ms |  41.99 ms | Stable (+2.9%)    |
| test_phenompv2_forward[batch_512-dur_2]      |  40.81 ms |  41.90 ms | Stable (+2.7%)    |
| test_phenompv2_forward[batch_512-dur_4]      |  43.76 ms |  44.78 ms | Stable (+2.3%)    |
| test_phenompv2_forward[batch_512-dur_8]      |  48.86 ms |  50.45 ms | Stable (+3.2%)    |
| test_td_generator_forward[batch_32-fmin_20]  |  52.82 ms |  54.43 ms | Stable (+3.0%)    |
| test_td_generator_forward[batch_32-fmin_30]  |  52.92 ms |  54.31 ms | Stable (+2.6%)    |
| test_td_generator_forward[batch_32-fmin_40]  |  52.78 ms |  54.37 ms | Stable (+3.0%)    |
| test_td_generator_forward[batch_128-fmin_20] |  64.60 ms |  67.60 ms | Stable (+4.6%)    |
| test_td_generator_forward[batch_128-fmin_30] |  64.69 ms |  65.98 ms | Stable (+2.0%)    |
| test_td_generator_forward[batch_128-fmin_40] |  64.73 ms |  66.00 ms | Stable (+2.0%)    |
| test_td_generator_forward[batch_512-fmin_20] | 107.99 ms |  95.78 ms | Improved (-11.3%) |
| test_td_generator_forward[batch_512-fmin_30] |  94.68 ms |  95.42 ms | Stable (+0.8%)    |
| test_td_generator_forward[batch_512-fmin_40] |  94.66 ms |  95.60 ms | Stable (+1.0%)    |
| test_phenomdeco_forward[batch_32-dur_2]      |       N/A |  37.38 ms | New               |
| test_phenomdeco_forward[batch_32-dur_4]      |       N/A |  37.47 ms | New               |
| test_phenomdeco_forward[batch_32-dur_8]      |       N/A |  37.20 ms | New               |
| test_phenomdeco_forward[batch_128-dur_2]     |       N/A |  37.44 ms | New               |
| test_phenomdeco_forward[batch_128-dur_4]     |       N/A |  37.31 ms | New               |
| test_phenomdeco_forward[batch_128-dur_8]     |       N/A |  45.08 ms | New               |
| test_phenomdeco_forward[batch_512-dur_2]     |       N/A |  45.16 ms | New               |
| test_phenomdeco_forward[batch_512-dur_4]     |       N/A |  61.80 ms | New               |
| test_phenomdeco_forward[batch_512-dur_8]     |       N/A | 113.08 ms | New               |
                                                                                            
                                                                                      
```

**Comment by deepchatterjeeligo** [comment]
Only the phenomdeco forward calls (A100):
```
---------------------------------------------------------------------------------------------- benchmark: 9 tests ----------------------------------------------------------------------------------------------
Name (time in ms)                                 Min                 Max                Mean            StdDev              Median               IQR            Outliers      OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_phenomdeco_forward[batch_32-dur_8]       35.8484 (1.0)       36.7997 (1.0)       36.3850 (1.0)      0.2553 (4.01)      36.4902 (1.01)     0.1613 (1.71)          6;6  27.4839 (1.0)          28           1
test_phenomdeco_forward[batch_32-dur_2]       36.1495 (1.01)      38.0487 (1.03)      36.4002 (1.00)     0.3618 (5.67)      36.2700 (1.0)      0.2733 (2.90)          1;1  27.4724 (1.00)         28           1
test_phenomdeco_forward[batch_128-dur_4]      36.1372 (1.01)      37.6262 (1.02)      36.6585 (1.01)     0.2945 (4.62)      36.6785 (1.01)     0.3192 (3.39)          6;1  27.2788 (0.99)         28           1
test_phenomdeco_forward[batch_32-dur_4]       36.2731 (1.01)      39.4239 (1.07)      36.7483 (1.01)     0.5619 (8.81)      36.7008 (1.01)     0.2954 (3.14)          1;1  27.2121 (0.99)         28           1
test_phenomdeco_forward[batch_128-dur_2]      36.2988 (1.01)      37.7281 (1.03)      36.7504 (1.01)     0.2687 (4.21)      36.7505 (1.01)     0.2827 (3.00)          5;1  27.2106 (0.99)         28           1
test_phenomdeco_forward[batch_128-dur_8]      44.3517 (1.24)      45.0207 (1.22)      44.6420 (1.23)     0.2126 (3.33)      44.5702 (1.23)     0.3461 (3.68)          7;0  22.4004 (0.82)         23           1
test_phenomdeco_forward[batch_512-dur_2]      44.3798 (1.24)      44.9600 (1.22)      44.6983 (1.23)     0.1555 (2.44)      44.7093 (1.23)     0.2181 (2.32)          8;0  22.3722 (0.81)         23           1
test_phenomdeco_forward[batch_512-dur_4]      60.9761 (1.70)      61.4919 (1.67)      61.2729 (1.68)     0.1601 (2.51)      61.2230 (1.69)     0.2477 (2.63)          6;0  16.3204 (0.59)         17           1
test_phenomdeco_forward[batch_512-dur_8]     113.5598 (3.17)     113.7300 (3.09)     113.6264 (3.12)     0.0637 (1.0)      113.6067 (3.13)     0.0941 (1.0)           3;0   8.8008 (0.32)          9           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

**Comment by deepchatterjeeligo** [comment]
@wbenoit26 can you take another look at this please? It is good from my end. Sorry about the mishap that happened with #240 

One issue that I see is the the DECO long duration large batch is slower compared to the PhenomD version.

```
| test_phenomd_forward[batch_512-dur_8]        |  38.36 ms |  37.77 ms | Stable (-1.5%) |
```
vs
```
| test_phenomdeco_forward[batch_512-dur_8]     |       N/A | 114.58 ms | New            |
```

But I am happy to get this in and investigate any optimizations as a second step

File: tests/waveforms/cbc/test_cbc_waveforms.py [hunk 1]
```diff
@@ -294,7 +294,7 @@ def test_phenom_d(
     lal_mask = (lal_freqs > params["f_min"]) & (lal_freqs < params["f_max"])
 
     lal_freqs = lal_freqs[lal_mask]
-    torch_freqs = torch.tensor(lal_freqs, dtype=FLOAT64)
+    torch_freqs = torch.tensor(lal_freqs, dtype=torch.float32)
```
Comment (wbenoit26) [inline tests/waveforms/cbc/test_cbc_waveforms.py]: Why was this changed from float64 to float32?

File: ml4gw/waveforms/cbc/phenom_deco.py [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+        eta2,
+        Seta,
+        chi1,
+        chi2,
+        chi12,
+        chi22,
+        xi,
+        distance,
+        c_eff,
+        fRD=None,  # used for passing ringdown frequency from phenom_p
```
Comment (wbenoit26) [inline ml4gw/waveforms/cbc/phenom_deco.py]: Back in this [commit](https://github.com/ML4GW/ml4gw/commit/54ea0db0927ad6bf04581ba36af4affde48b78e9), we changed how `fRD` and `fDM` are handled. It would be good to have it done that here as well.

**Review by wbenoit26 — CHANGES_REQUESTED** [review]
Couple of minor things, overall looks good.

File: ml4gw/waveforms/cbc/phenom_deco.py (hunk 2 above)
Comment (deepchatterjeeligo) [reply ml4gw/waveforms/cbc/phenom_deco.py]: good point. Removed those if conditions and optional None for fRD and fMD. Should be consistent with the PhenomD methods now.

File: tests/waveforms/cbc/test_cbc_waveforms.py (hunk 1 above)
Comment (deepchatterjeeligo) [reply tests/waveforms/cbc/test_cbc_waveforms.py]: not sure, reverted it to float64, and re ran the tests and benchmarks. We get similar timings.

**Review by deepchatterjeeligo — COMMENTED** [review]
@wbenoit26 hopefully one last look.

**Comment by deepchatterjeeligo** [comment]
Oh and BTW the `test_td_generator_forward` benchmarks timing in my comment above says "Improved", but that may have been something sporadic. I reran two more times and I get "Stable", which makes sense as that is not supposed to change much.

```
| test_td_generator_forward[batch_512-fmin_20] | 107.99 ms | 109.60 ms | Stable (+1.5%) |
```

**Review by wbenoit26 — APPROVED** [review]
Looks good to me.

**Comment by wbenoit26** [comment]
> Oh and BTW the `test_td_generator_forward` benchmarks timing in my comment above says "Improved", but that may have been something sporadic. I reran two more times and I get "Stable", which makes sense as that is not supposed to change much.
> 
> ```
> | test_td_generator_forward[batch_512-fmin_20] | 107.99 ms | 109.60 ms | Stable (+1.5%) |
> ```

Ah, okay, that makes sense. I wonder if increasing the number of warm-up iterations would help, or if it's just something we'll need to be aware of.

**Comment by deepchatterjeeligo** [comment]
I increased the number of warmups from 5 to 50 and redid the benchmarks, and the `test_td_generator_forward` still turns out to be `Stable` so I think it was something one-off. Just something to be mindful of.
